R/breslow.R

```diff
@@ -0,0 +1,156 @@
+#'
+#' @description
+#' Helper function to compose a survival distribution from the relative risk
+#' predictions (`lp`) of a **proportional hazards** model (e.g. a Cox-type model).
+#'
+#' @param times (`numeric()`)\cr Vector of times (train set).
+#' @param status (`numeric()`)\cr Vector of status indicators (train set).
+#' For each observation in the train set, this should be 0 (alive/censored) or
+#' 1 (dead).
+#' @param lp_train (`numeric()`)\cr Vector of linear predictors (train set).
+#' These are the relative score predictions (\eqn{lp = \hat{\beta}X_{train}})
+#' from a proportional hazards model on the train set.
+#' @param lp_test (`numeric()`)\cr Vector of linear predictors (test set).
+#' These are the relative score predictions (\eqn{lp = \hat{\beta}X_{test}})
+#' from a proportional hazards model on the test set.
+#' @param eval_times (`numeric()`)\cr Vector of times to compute survival
+#' probabilities. If `NULL` (default), the unique and sorted `times` from the
+#' train set will be used, otherwise the unique and sorted `eval_times`.
+#'
+#' @details
+#' We estimate the survival probability of individual \eqn{i} (from the test set),
+#' at time point \eqn{t} as follows:
+#' \deqn{S_i(t) = e^{-H_i(t)} = e^{-\hat{H}_0(t) \times e^{lp_i}}}
+#'
+#' where:
+#' - \eqn{H_i(t)} is the cumulative hazard function for individual \eqn{i}
+#' - \eqn{\hat{H}_0(t)} is Breslow's estimator for the cumulative baseline
+#' hazard. Estimation requires the training set's `times` and `status` as well
+#' the risk predictions (`lp_train`), see more details in [.cbhaz_breslow])
+#' - \eqn{lp_i} is the risk prediction (linear predictor) of individual \eqn{i}
+#' on the test set.
+#'
+#' @return a survival probability `matrix` (obs x times). Number of columns is
+#' equal to `eval_times` and number of rows is equal to the number of test
+#' observations (i.e. the length of the `lp_test` vector).
+#'
+#' @references
+#' `r format_bib("cox_1972", "lin_2007")`
+#'
+#' @examples
+#' task = tsk("rats")
+#' part = partition(task, ratio = 0.8)
+#'
+#' learner = lrn("surv.coxph")
+#' learner$train(task, part$train)
+#' p_train = learner$predict(task, part$train)
+#' p_test  = learner$predict(task, part$test)
+#'
+#' surv = surv_breslow(times = task$times(part$train), status = task$status(part$train),
+#'                     lp_train = p_train$lp, lp_test = p_test$lp)
+#' head(surv)
+#' @export
+surv_breslow = function(times, status, lp_train, lp_test, eval_times = NULL) {
+  assert_numeric(times, null.ok = FALSE)
+  assert_numeric(status, null.ok = FALSE)
+  assert_numeric(lp_train, null.ok = FALSE)
+  assert_true(length(times) == length(status))
+  assert_true(length(times) == length(lp_train))
+  assert_numeric(lp_test, null.ok = FALSE)
+  assert_numeric(eval_times, null.ok = TRUE)
+
+  base_haz = .cbhaz_breslow(times = times, status = status, lp = lp_train,
+                            eval_times = eval_times)
+  surv = exp(exp(lp_test) %*% -t(base_haz))
```

Comment:
I don't understand this. .cbhaz_breslow should return the cumulative hazard but you've got a further computation here, I'd expect this line to be `exp(-base_haz)` and that's it

Reply:
Sorry different point. My point is that the end of .cbhaz_breslow should be `exp(lp_test) %*% t(base_haz)` as *that* is the cumulative hazard by Breslow. Then you run like 63 here and after that you run `exp(-base_haz)` to convert to survival. Otherwise `.cbhaz_breslow` is doing a different thing.

Reply:
discussed and solved, will generalize the function to have two output types